### ISSUE TITLE
Question about logging support ?

### ISSUE THREAD

**Issue by Mostrapotski** [joffrey-bion/krossbow#66]
Hello, 

When I send messages, I never receive any response. 
Logging everything going out or in would be a good place for me to start debugging.

Maybe i overlooked but i don't see how to do it. 
I must have overlooked it, because I think log are useful even for writing krossbow.

**Comment by joffrey-bion** [comment]
Hi @Mostrapotski, thanks for your issue. I actually haven't added any logs so far in the library itself, because I used it mostly in a web browser (where I can see all websocket frames) and in unit tests.

I have actually never looked into Kotlin logging in details. Do you know if there is a Kotlin multiplatform equivalent for SLF4J for instance? 
I will definitely investigate that if you don't know off the top of your head.

**Comment by Mostrapotski** [comment]
Well, i don't know :-1: 

But i believe at some point logging will be very platform or even use case dependent. 
Some users might want to provide their own logging output, so maybe we can create not a multiplatform logging tool (some already exists on github) but rather a log interface, like a log factory, where krossbow will push logs (and the responsibility to output them). 

Maybe we can add a default basic impl for this, so it's usage out of the box ?

**Comment by joffrey-bion** [comment]
@Mostrapotski that's exactly what SLF4J is, a common logging interface for which you can plug any logging implementation. And `kotlin-logging` is built on top of SLF4J in its JVM version, so I think I'll give this one a shot.

**Comment by joffrey-bion** [comment]
On second thought, I wonder in which place logging would be useful from inside Krossbow. If I add debug logs for all incoming/outgoing messages, this might incur a performance cost that's significant enough for low-latency applications that rely on websockets to be fast.

In order to find your actual current problem, are you sure there aren't any uncaught exceptions or things like that, which you can already see in the logs? Most errors in Krossbow translate into exceptions that are bubbled up to the consumer's coroutines. Could you please give a bit of context about your issue?

**Comment by Mostrapotski** [comment]
I hear the performance issue, but maybe we can do it for debug only, either with a config flag or something from the platform. e.g on android, you can build your app in release or in debug. 
For HTTP, i do something like this:
```
          if (BuildConfig.BUILD_TYPE == "debug") {
                addInterceptor(HttpLoggingInterceptor().apply {
                    level = HttpLoggingInterceptor.Level.BODY
                })
                addInterceptor(CurlLoggerInterceptor())
            }
```
Only cost is a bool check. 
If looking at the logs from the browser was helpful for you, then it will be on every platform. 

As you proposed, I can log by myself, but krossbow core is more "low level". If I were a bad developer I could log that I sent a message, even if the socket is not connected, the lower level it is, the more we can trust the logs. 

Concerning my problem, the server is excepting rather complex messages, I just sent "toto" and excepted "something" (I subscribed to `/user/queue/error` which is specified in my server documentation), but nothing. Oo exceptions nor crashes. I will try to access the server logs tomorrow.

**Comment by joffrey-bion** [comment]
> For HTTP, i do something like this

Yes this looks like the instrumentation hook I was mentioning. This is something I can definitely consider, as it can be useful for different things.

> If looking at the logs from the browser was helpful for you, then it will be on every platform.

This was helpful for me in order to debug Krossbow itself mostly, that's why I didn't expect this to be needed from end users. That being said, I definitely get that if you want to just "log everything that goes in or out", you currently have no way with Krossbow to do it, and you would have to go to every place in your code where you send a message, which is pretty annoying. That's why I'm considering adding some instrumentation hook (like the interceptor mechanism).

> As you proposed, I can log by myself, but krossbow core is more "low level". If I were a bad developer I could log that I sent a message, even if the socket is not connected, the lower level it is, the more we can trust the logs.

That's why I am mentioning that Krossbow should bubble exceptions. If you use a `session.send()` method and log a message *afterwards*, your log should only be written if the frame was successfully sent over the websocket connection (otherwise krossbow would fail with an exception).
That's why I'm saying that unless you're actually trying to debug Krossbow itself (which may very well have a bug!) you shouldn't need "insider" logs.

> Concerning my problem, the server is excepting rather complex messages, I just sent "toto" and excepted "something" (I subscribed to /user/queue/error which is specified in my server documentation), but nothing. No exceptions nor crashes. I will try to access the server logs tomorrow.

Which server are you using? Spring? I have been using Spring for a while, and my main problem when this happened was missing heart beats, but you should see exceptions if that were the problem (and this only happens if you enabled heart beats of course). Spring really is lenient in when it sends heart beats (sometimes a few seconds late!). So Krossbow tends to consider it a connection drop. This is why the heart beat tolerance options came from.

So as a summary, I believe you shouldn't need internal logging or instrumentation to investigate this particular problem, but it may be useful for global instrumentation/interception of messages in general. So I'll work on this anyway.
This is of course unless there is a bug in Krossbow, in which case it might be painful to find out from your side, but then it will lead to a fix of the issue 😄 .

**Comment by joffrey-bion** [comment]
I'm closing this issue as instrumentation is available as of version 0.32.0 (doc will follow, but basically you can provide a `KrossbowInstrumentation` implementation in the `StompConfig`).

There doesn't seem to be a need for logging at the moment, as errors are already handled as exceptions and other events can be known from the user's perspective by observing when the suspending functions return.

If you're lacking specific features in instrumentation, please open other issues.
